show-coordinates-in-multiple-formats/src/main/java/com/esri/arcgismaps/sample/showcoordinatesinmultipleformats/components/MapViewModel.kt

```diff
@@ -0,0 +1,195 @@
+/* Copyright 2023 Esri
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.esri.arcgismaps.sample.showcoordinatesinmultipleformats.components
+
+import android.app.Application
+import androidx.compose.runtime.mutableStateOf
+import androidx.lifecycle.AndroidViewModel
+import com.arcgismaps.Color
+import com.arcgismaps.geometry.CoordinateFormatter
+import com.arcgismaps.geometry.LatitudeLongitudeFormat
+import com.arcgismaps.geometry.Point
+import com.arcgismaps.geometry.SpatialReference
+import com.arcgismaps.geometry.UtmConversionMode
+import com.arcgismaps.mapping.ArcGISMap
+import com.arcgismaps.mapping.Basemap
+import com.arcgismaps.mapping.BasemapStyle
+import com.arcgismaps.mapping.layers.ArcGISTiledLayer
+import com.arcgismaps.mapping.symbology.SimpleMarkerSymbol
+import com.arcgismaps.mapping.symbology.SimpleMarkerSymbolStyle
+import com.arcgismaps.mapping.view.Graphic
+import com.arcgismaps.mapping.view.GraphicsOverlay
+import com.esri.arcgismaps.sample.showcoordinatesinmultipleformats.R
+import kotlinx.coroutines.flow.MutableStateFlow
+
+class MapViewModel(application: Application) : AndroidViewModel(application) {
+    // set the MapView mutable stateflow
+    val mapViewState = MutableStateFlow(MapViewState())
+
+    var decimalDegrees = mutableStateOf("")
+    var degreesMinutesSeconds = mutableStateOf("")
+    var utm = mutableStateOf("")
+    var usng = mutableStateOf("")
+
+    // set up a graphic to indicate where the coordinates relate to, with an initial location
+    private val initialPoint by lazy {
+        Point(0.0, 0.0, SpatialReference.wgs84())
+    }
+    private val coordinateLocation by lazy {
+        Graphic(
+            geometry = initialPoint,
+            symbol = SimpleMarkerSymbol(
+                style = SimpleMarkerSymbolStyle.Cross,
+                color = Color.fromRgba(255, 255, 0, 255),
+                size = 20f
+            )
+        )
+    }
+
+    init {
+        // create a map that has the WGS 84 coordinate system and set this into the map
+        val basemapLayer = ArcGISTiledLayer(application.getString(R.string.basemap_url))
+        val map = ArcGISMap(Basemap(basemapLayer))
+        mapViewState.value.apply {
+            arcGISMap = map
+            graphicsOverlay.graphics.add(coordinateLocation)
+        }
```

Comment:
I followed this [codelab example](https://developer.android.com/codelabs/basic-android-kotlin-compose-viewmodel-and-state#4) as a guide to set up the MapView state as a `MutableStateFlow`. To me it seems like the approach of using the `MapViewState` as a mutable state (`val mapViewState by mutableStateOf(MapViewState())`) could be a viable option. 

I made some changes, and I am curious to know if you would recommend this approach moving forward or should we stick to the code lab example of using a `MutableStateFlow`?

Reply:
I think either `MutableState` or `MutableStateFlow` is okay in this scenario. Actually I had a discussion about this with Gunther yesterday regarding toolkit, and he was leaning towards not using `State` or `MutableState` in `ViewModel`s at all. I think in this case it's okay because the sample VM will probably not be copied over wholesale into somebody's app, we don't need to worry about people potentially using classic Android views and therefore needing the UI-agnostic `StateFlow`

My previous comment was actually about the class `MapViewState` though. `MapViewState` has two public, mutable properties, and the views that observe it only refer to the single instance of `MapViewState`. So I would not expect `MapViewModel.mapViewState` to be a `MutableState` because only one instance is ever used, instead I would expect `MapViewState.arcGISMap` and `MapViewState.graphicsOverlay` to be `MutableState`s. Alternatively, I would expect `MapViewState.arcGISMap and MapViewState.graphicsOverlay` to be immutable `val`s and `MapViewModel.mapViewState` to remain a `MutableState`, so that any changes to `MapViewModel.mapViewState` would be observed.

I don't think I wrote that well so here's the two versions I'm describing:

1.
```kotlin 
// Note this can't be a data class if using delegated properties
class MapViewState {
    var arcGISMap: ArcGISMap by mutableStateOf(ArcGISMap(BasemapStyle.ArcGISNavigationNight)),
    var graphicsOverlay: GraphicsOverlay by mutableStateOf(GraphicsOverlay())
}

class MapViewModel(application: Application) : AndroidViewModel(application) {
    // set the MapView state
    val mapViewState = MapViewState()
```

2.
```kotlin
data class MapViewState(
    val arcGISMap: ArcGISMap = ArcGISMap(BasemapStyle.ArcGISNavigationNight),
    val graphicsOverlay: GraphicsOverlay = GraphicsOverlay()
)

class MapViewModel(application: Application) : AndroidViewModel(application) {
    // set the MapView state
    val mapViewState by mutableStateOf(MapViewState())
```

As it is currently written, setting `MapViewModel.mapViewState.arcGISMap = ArcGISMap(BasemapStyle.ArcGISColoredPencil)` won't do anything anything. I tested this by changing the code in `onSingleTapConfirmed` to set the map, and nothing happens, and with the debugger the update block line that changes the map never gets called. That's because neither the property `MapViewModel.mapViewState` is changing, nor does `MapViewModel.mapViewState.arcGISMap` participate in the state system, so Compose doesn't know how to observe it.

In either case 1 or 2 I think that `ComposeMapView` stays exactly as is, the only difference is what things are changing--is it the state object or is it the properties on the state object? I think the biggest practical difference is how exactly the update block of the `AndroidView` will run if only one property (let's say `arcGISMap`) changes. In option 1, the update block will only run the line that updates the map. But in option 2, I think that the entire update block will run because the whole object is updated (maybe there are some optimizations under the hood). I don't think there's a real reason to do one or the other though given how small this class is. I wrote a quick test of both using the same method above and they both trigger changes to the map.

Reply:
@hud10837 Thanks, this was easy to understand with above explanation and examples. +1 for option 1.